vignettes/stages.Rmd

```diff
@@ -0,0 +1,103 @@
+
+The lifecycle stages can apply to packages, functions, function arguments, and even specific values of a functions.
+However, you'll mostly see them used to label individual functions, so that's the language we use below.
+
+## Stable
+
+The default development stage is ![stable](figures/lifecycle-stable.svg){style="vertical-align:middle"}.
+A function is considered stable when the author is happy with its interface, doesn't see major issues, and is happy to share it with the world.
+Because this stage is the default, functions will only be given a stable badge if there's some specific need to draw attention to their status.
+
+Stability is defined in terms of breaking changes.
+A **breaking change** is a change that breaks code that uses the function as expected.
+In general, breaking changes reduce the set of code that works without error, either by removing a function, removing a function argument, or decreasing the set of valid inputs.
+Breaking changes also include changes to output type.
+If you imagine all the possible inputs to a function that return a result (and not an error), a breaking change makes the set smaller.
+
+Not all changes that cause your function to stop working are breaking changes.
+For example, you might have accidentally relied on a bug.
+When the bug is fixed, your code breaks, but this is not a breaking change.
+A good way of making your code more robust to this sort of behaviour change is to only use a function for its explicitly intended effects.
+For example, using `c()` to concatenate two vectors will not put your code at risk because it clearly is the intended usage of this function.
+On the other hand, using `c()` just for the side effect of removing attributes is probably not a good idea.
+If you use `c(factor("foo"))` to retrieve the underlying integers of the factor levels, your code is at risk of breaking if `c()` ever implements a factor method.
+
+Stable functions come with two promises related to breaking changes:
+
+-   Breaking changes will be avoided where possible.
+    We'll only make breaking changes if we consider the long term benefit of the change to be greater than short term pain of changing existing code.
+
+-   If a breaking change is needed, it will occur gradually, through the deprecation process described next.
+    This gives you plenty of time to adjust your code before it starts generating errors.
+
+## Deprecated
+
+A ![Deprecated](figures/lifecycle-deprecated.svg){style="vertical-align:middle"} function has a better alternative available and is scheduled for removal.
+When you call a deprecated function, you get a warning telling you what to use instead.
+For example, take `tibble::as_data_frame()`:
+
+```{r, eval = FALSE}
+df <- tibble::as_data_frame(x = 1)
+#> Warning message:
+#> `data_frame()` is deprecated as of tibble 1.1.0.
```

Comment:
It's confusing that the call is to `as_data_frame()`, but the warning is about `data_frame()`. Pick a less funky example?

Reply:
Oh that's just a think-o.